### PR TITLE
(G)DMA AES working mode support

### PR DESCRIPTION
**Pull request by JurajSadel**
## Thank you!

Thank you for your contribution.
Please make sure that your submission includes the following:

### Must

- [x] The code compiles without `errors` or `warnings`.
- [x] All examples work.
- [x] `cargo fmt` was run.
- [x] Your changes were added to the `CHANGELOG.md` in the proper section.
- [x] You updated existing examples or added examples (if applicable).
- [x] Added examples are checked in CI

### Nice to have

- [x] You add a description of your work to this PR.
- [x] You added proper docs for your newly added features and code.

This doesn't add support for ESP32 and ESP32S2. I will create a separate PR for these two chips. Also, I left a few `TODO`s mostly for ESP32 and I will address them in the next PR. I will add `async` support next and after that, I will add DMA for ESP32 and S2.

### CODE REVIEW

File: esp-hal-common/src/dma/gdma.rs [hunk 1]
```diff
@@ -609,6 +609,7 @@ macro_rules! impl_channel {
             impl I2s1Peripheral for [<SuitablePeripheral $num>] {}
             #[cfg(parl_io)]
             impl ParlIoPeripheral for [<SuitablePeripheral $num>] {}
+            impl AesPeripheral for [<SuitablePeripheral $num>] {}
```
Comment (bjoernQ) [inline esp-hal-common/src/dma/gdma.rs:612]: feature gate this?

File: esp-hal-common/src/dma/mod.rs [hunk 2]
```diff
@@ -135,7 +135,7 @@ pub enum DmaPeripheral {
     I2s1   = 4,
     #[cfg(esp32s3)]
     LcdCam = 5,
-    #[cfg(any(esp32c3, esp32c6, esp32h2, esp32s3))]
+    #[cfg(not(any(esp32, esp32c2)))]
```
Comment (bjoernQ) [inline esp-hal-common/src/dma/mod.rs:138]: why is there esp32 but not esp32s2?
probably both of them or none of them?

File: esp32-hal/examples/aes.rs [hunk 3]
```diff
@@ -35,19 +35,17 @@ fn main() -> ! {
     let mut block_buf = [0_u8; 16];
     block_buf[..plaintext.len()].copy_from_slice(plaintext);
 
-    let key = Key::<Aes128>::from(&keybuf);
-    let mut cipher = Cipher::new(&mut aes, &key);
     let mut block = block_buf.clone();
     let pre_hw_encrypt = xtensa_lx::timer::get_cycle_count();
-    cipher.encrypt_block(&mut block);
+    aes.start_operation(&mut block, Mode::Encryption128, &keybuf);
```
Comment (bjoernQ) [inline esp32-hal/examples/aes.rs:40]: `start_operation` is a bit misleading for a blocking operation that is finished when it returns
`start_xxx` would imply there is a way to finish the operation later

File: esp32c3-hal/examples/aes_dma.rs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    keybuf[..keytext.len()].copy_from_slice(keytext);
+
+    // create an array with aes block size
+    let mut block_buf = [0_u8; 16];
+    block_buf[..plaintext.len()].copy_from_slice(plaintext);
+
+    let hw_encrypted = buffer3();
+    let pre_hw_encrypt = SystemTimer::now();
+    let transfer = aes
+        .dma_transfer(
```
Comment (bjoernQ) [inline esp32c3-hal/examples/aes_dma.rs:60]: not sure if the name is really good - while it does a DMA transfer it's probably not the interesting part for the public API

**Comment by bjoernQ** [comment]
Some bikeshedding but generally looks good

**Comment by JurajSadel** [comment]
Thank you for the review @bjoernQ! I addressed your comments.

File: esp-hal-common/src/aes/mod.rs [hunk 5]
```diff
@@ -142,88 +220,320 @@ pub struct Aes192;
...
+        fn with_dma(self, mut channel: Channel<'d, C>) -> AesDma<'d, C> {
+            channel.tx.init_channel(); // no need to call this for both, TX and RX
+
+            AesDma { aes: self, channel }
+        }
     }
 
-    fn is_idle(&mut self) -> bool {
-        self.aes.read_idle()
+    /// An in-proress DMA transfer
```
Comment (bjoernQ) [inline esp-hal-common/src/aes/mod.rs:292]: typo "in-progress"

**Review by bjoernQ — APPROVED** [review]
LGTM - Thanks for working on this

**Review by MabezDev — APPROVED** [review]
Nice, thanks!

**Review by jessebraham — APPROVED** [review]
LGTM, thanks for this!

**Pull request merged**